File: sections/40_data_structures_mixed_contest/problems/c_mo_distinct_queries/solution.cpp
#include <bits/stdc++.h>
using namespace std;

struct Query {
    int l, r, id;
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n, q;
    if (!(cin >> n >> q)) return 0;
    vector<int> a(n), vals;
    for (int &x : a) {
        cin >> x;
        vals.push_back(x);
    }
    sort(vals.begin(), vals.end());
    vals.erase(unique(vals.begin(), vals.end()), vals.end());
    for (int &x : a) x = int(lower_bound(vals.begin(), vals.end(), x) - vals.begin());
    vector<Query> queries(q);
    for (int i = 0; i < q; i++) {
        cin >> queries[i].l >> queries[i].r;
        --queries[i].l; --queries[i].r;
        queries[i].id = i;
    }
    int block = max(1, (int)sqrt(n));
    sort(queries.begin(), queries.end(), [&](const Query& A, const Query& B) {
        int ba = A.l / block, bb = B.l / block;
        if (ba != bb) return ba < bb;
        return (ba & 1) ? A.r > B.r : A.r < B.r;
    });
    vector<int> freq(vals.size(), 0), ans(q);
    int distinct = 0, l = 0, r = -1;
    auto add = [&](int pos) {
        if (freq[a[pos]]++ == 0) distinct++;
    };
    auto remove = [&](int pos) {
        if (--freq[a[pos]] == 0) distinct--;
    };
    for (auto qu : queries) {
        while (r < qu.r) add(++r);
        while (r > qu.r) remove(r--);
        while (l < qu.l) remove(l++);
        while (l > qu.l) add(--l);
        ans[qu.id] = distinct;
    }
    for (int x : ans) cout << x << '\n';
    return 0;
}

File: sections/40_data_structures_mixed_contest/problems/c_mo_distinct_queries/solution.py
import sys
from collections import defaultdict

def main():
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    n, q = data[0], data[1]
    a = data[2:2 + n]
    idx = 2 + n
    block = max(1, int(n ** 0.5))
    queries = []
    for qi in range(q):
        l, r = data[idx] - 1, data[idx + 1] - 1
        idx += 2
        queries.append((l, r, qi))
    queries.sort(key=lambda x: (x[0] // block, x[1] if (x[0] // block) % 2 == 0 else -x[1]))
    freq = defaultdict(int)
    ans = [0] * q
    distinct = 0
    left, right = 0, -1
    for l, r, qi in queries:
        while right < r:
            right += 1
            freq[a[right]] += 1
            if freq[a[right]] == 1:
                distinct += 1
        while right > r:
            freq[a[right]] -= 1
            if freq[a[right]] == 0:
                distinct -= 1
            right -= 1
        while left < l:
            freq[a[left]] -= 1
            if freq[a[left]] == 0:
                distinct -= 1
            left += 1
        while left > l:
            left -= 1
            freq[a[left]] += 1
            if freq[a[left]] == 1:
                distinct += 1
        ans[qi] = distinct
    print("\n".join(map(str, ans)))

if __name__ == "__main__":
    main()
